flocker/control/_diffing.py

```diff
@@ -111,9 +262,23 @@ class Diff(PClass):
     changes = pvector_field(object)
 
     def apply(self, obj):
+        proxy = _EvolverProxy(original=obj)
         for c in self.changes:
-            obj = c.apply(obj)
-        return obj
+            if len(c.path) > 0:
+                proxy = c.apply(proxy)
+            else:
+                assert type(c) is _Set
+                proxy = _EvolverProxy(original=c.value)
+        try:
+            return proxy.commit()
+        except:
+            # Imported here to avoid circular dependencies.
+            from ._persistence import wire_encode
```

Comment:
I had a go at re-organizing the modules in #2858 but it needs more thought.
For now, I'm going to leave it.